---
layout: post
title: Your toolbox
---

A common question from computational newcomers is, "Where should I start?"

## Galaxy

If you have a question that is commonly encountered by other people doing biological research, your first step should be to take a look at Galaxy.
Galaxy is many things, but from most users' perspective it's a web application for analyzing large amounts of biomedical data.
There are public Galaxy instances, like [Galaxy main](http://usegalaxy.org), and private ones, like our own [shapley.fhcrc.org](http://shapley.fhcrc.org), which runs on the same infrastructure as the rest of the Hutch's scientific computing services.
Regardless of where the instance is, Galaxy makes available a large number of standard analysis programs (cufflinks, MACS, EMBOSS, etc.) wrapped in a much friendlier interface than the programs themselves are typically known for.
In addition to the tools included with every Galaxy installation, there are repositories called *tool sheds* where tools developed by the community are shared.
The [main Galaxy Tool Shed](https://toolshed.g2.bx.psu.edu) is easy to search through, if a little slow, and contains 2490 tools at time of writing.
Most tools can be easily installed onto a private Galaxy instance like ours in a few minutes by an administrator.

Galaxy really shines with its ability to save and share reproducible workflows.
In Galaxy parlance, a *workflow* is simply a recipe for analyzing data that typically consists of multiple steps, multiple tools, etc. Taking an example from some of our microbiome work, a single workflow could encapsulate quality filtering raw 454 reads stored in FASTA format, computing an alignment, mapping the aligned sequences, and then computing diversity statistics.
Sure, creating workflows like the above reduce duplication of effort, but arguably more importantly it enhances reproducibility.
For each workflow, Galaxy saves the *exact* tool versions used when the workflow was created, and when possible, will always use those specific versions when the workflow is run again.
The workflows published by the Galaxy main community are publicly searchable [here](https://usegalaxy.org/workflow/list_published), but your own workflows can also be shared privately with certain users or exported as a file and saved to your computer.

For more information on what Galaxy is and how to use it, visit the Galaxy project's [learning annex](https://wiki.galaxyproject.org).

## R

If your data is in tabular format and you want to learn from it, the answer is almost certainly to use R.
R is a fantastic tool to munge, summarize, extract, plot, statistically analyze, and otherwise learn from your data.
If you're a developer but you haven't used a vectorized language like Fortran 95 or MATLAB before, avoiding loops might seem difficult at first, but it's usually well worth the effort in terms of code clarity and efficiency.
The offical introduction to R is available at [r-project.org](http://cran.r-project.org/doc/manuals/R-intro.html).

Another great feature of R is [CRAN](http://cran.us.r-project.org), the Comprehensive R Archive Network.
CRAN is sort of like Galaxy's tool sheds, and is one of the first places to look if you want to see if a well-developed solution to your problem already exists.
Here's a poignant anecdote I like from Martin Morgan, about one of his first experiences with R:

> Iâ€™d initially written a C program that was literally thousands of lines of code. After a while, I decided to try programming in R; since the facilities for data input and optimization were already there, I was able to do it in just six lines of R code.

R is also well-known for its ability to produce beautiful visualizations using packages like [ggplot2](http://docs.ggplot2.org). If data visualization is in your future, R might be a good bet.

## Python

If your data is not in tabular format, or if you want to do some analysis *and* run some external programs, Python is a gorgeous choice.
Python is a solid and popular general-purpose language that many developers fall back to when there isn't a compelling reason to use a more specialized tool.
It is also arguably one of the best-documented modern languages; tutorials and references can all be found on [docs.python.org](http://docs.python.org).
(Just make sure you're reading the docs for the right version of python; there were major changes to the language between versions 2 and 3.)

Like R, Python has a central repository with thousands of packages you can "import" into your own programs: e.g., [BioPython](http://biopython.org) for biological data and analysis; [pymc](http://pymc-devs.github.io/pymc) for Bayesian statistical modeling; or [Sage](http://www.sagemath.org) for "elementary and advanced, pure and applied" mathematics.

![xkcd comic about Python](http://imgs.xkcd.com/comics/python.png)

Of course, there are plenty of more mundane, utilitarian packages that are nonetheless invaluable for day-to-day things.

## Shell scripting

If you only want to run command line programs, a shell script is the ticket.
Shell scripts are the command-line equivalent of Galaxy's workflows, and behave almost exactly as if you were sitting at the keyboard typing the commands in manually.
In fact, getting started with shell scripting is about that easy -- open up an editor, type one command per line, and then have your computer run it.
We'll be covering shell scripting a bit later, but the Linux Documentation Project's [Bash Guide for Beginners](http://www.tldp.org/LDP/Bash-Beginners-Guide/html/Bash-Beginners-Guide.html) is a good place to start.

## What if I get stuck?

We all do! The tools mentioned here all have active online communities, and sometimes the community is the best part.
[Biostars](http://www.biostars.org) is devoted specifically to Galaxy users, while [Stack Overflow](http://stackoverflow.com) is a great resource for questions about Python, R, shell scripts, or just about any other programming language.
